src/topology/uniform_space/uniform_convergence_topology.lean

```diff
@@ -588,6 +590,155 @@ begin
+
+/-- For `𝔖 𝔗 : set (set α)`, `generate_same 𝔖 𝔗` expresses the fact that `𝔖` and `𝔗` generate the
+same *noncovering* bornology on `α` (i.e ideal of `set α`). We express this by saying that the
+pointwise complements of `𝔖` and `𝔗` generate the same filter. See `generate_same_iff` for
+a more concrete characterization. -/
+def generate_same (𝔖 𝔗 : set (set α)) : Prop :=
+generate (compl ⁻¹' 𝔖) = generate (compl ⁻¹' 𝔗)
+
+/-- For every `𝔖`, there is a nonempty and directed `𝔗` which satisfies `generate_same 𝔗`. This
+is essentially saying that any *noncovering* bornology admits a basis (in the dual sense of
+`filter.has_basis`). We could even ask `𝔗` to be downwards-closed (i.e a full-blown *noncovering*
+bornology), but it's not useful in practice. -/
+lemma exists_generate_same_directed (𝔖 : set (set α)) :
+  ∃ 𝔗 : set (set α), 𝔗.nonempty ∧ directed_on (⊆) 𝔗 ∧ generate_same 𝔖 𝔗 :=
+begin
+  -- The proof is simple: the bornology generated by `𝔖` works.
+  -- As usual, we take complements to go back to filters
+  refine ⟨{s | sᶜ ∈ filter.generate (compl ⁻¹' 𝔖)}, ⟨∅, show ∅ᶜ ∈ _, from _⟩,
+    directed_on_of_sup_mem (λ s t hs ht, show (s ∪ t)ᶜ ∈ _, from _), _⟩,
+  { rw compl_empty,
+    exact univ_mem },
+  { rw compl_union,
+    exact inter_mem hs ht },
+  { change _ = generate ((compl ∘ compl) ⁻¹' (generate _).sets),
+    rw [compl_comp_compl, preimage_id, (gi_generate α).l_u_eq] }
+end
+
+/-- Concrete characterization of `generate_same`. -/
+lemma generate_same_iff {𝔖 𝔗 : set (set α)} :
+  generate_same 𝔖 𝔗 ↔
+    (∀ s ∈ 𝔖, ∃ T ⊆ 𝔗, T.finite ∧ s ⊆ ⋃₀ T) ∧ (∀ t ∈ 𝔗, ∃ S ⊆ 𝔖, S.finite ∧ t ⊆ ⋃₀ S) :=
+begin
+  -- Like above, we go back to filters to apply `mem_generate_iff`.
+  -- Eventually, this should be a direct consequence of the dual lemmas for bornologies,
+  -- which would make the proof much simpler because we wouldn't have to take care of complements
+  symmetry,
+  rw [generate_same, and_comm, le_antisymm_iff, sets_iff_generate, sets_iff_generate, subset_def],
+  congrm (_ ∧ _);
+  rw [(@compl_surjective (set α) _).forall];
+  refine forall₂_congr (λ x hx, _);
+  rw [filter.mem_sets, mem_generate_iff];
+  { split; rintros ⟨Y, hY𝔖, hYf, hxY⟩,
+    { refine ⟨compl ⁻¹' Y, preimage_mono hY𝔖, hYf.preimage (compl_injective.inj_on _), _⟩,
+      rwa [compl_subset_comm, compl_sUnion, ← preimage_compl_eq_image_compl] at hxY },
+    { refine ⟨compl '' Y, image_subset_iff.mpr hY𝔖, hYf.image _, _⟩,
+      rwa [← compl_sInter, compl_subset_compl] } }
+end
+
+/-- Key part of proving `uniform_on_fun_filter_generate`: given any uniform structure `u` on
+    `α → β`, the set of all `S : set α` for which the restriction map from `(α → β, u)` → `S →ᵤ β`
+    is uniformly continuous is a *noncovering* bornology on `α`.
+    As usual, we take complements to express that in terms of filters. -/
+def co_converging_sets (u : uniform_space (α → β)) :
+  filter α :=
+{ sets := {S | @uniform_continuous (α → β) ((Sᶜ : set α) →ᵤ β) u _ (restrict Sᶜ)},
```

Comment:
This line is really terrifying. I guess you thought about trying to express this in terms of other operations?